Source/WTF/wtf/WeakListHashSet.h

```diff
@@ -0,0 +1,306 @@
+        return false;
+    }
+
+    template <typename U>
+    AddResult add(const U& value)
+    {
+        amortizedCleanupIfNeeded();
+        return m_set.add(*static_cast<const T&>(value).weakPtrFactory().template createWeakPtr<T>(const_cast<U&>(value), assertionsPolicy).m_impl);
+    }
+
+    template <typename U>
+    AddResult appendOrMoveToLast(const U& value)
+    {
+        amortizedCleanupIfNeeded();
+        return m_set.appendOrMoveToLast(*static_cast<const T&>(value).weakPtrFactory().template createWeakPtr<T>(const_cast<U&>(value), assertionsPolicy).m_impl);
+    }
+
+    template <typename U>
+    AddResult prependOrMoveToFirst(const U& value)
+    {
+        amortizedCleanupIfNeeded();
+        return m_set.prependOrMoveToFirst(*static_cast<const T&>(value).weakPtrFactory().template createWeakPtr<T>(const_cast<U&>(value), assertionsPolicy).m_impl);
+    }
+
+    template <typename U, typename V>
+    AddResult insertBefore(const U& beforeValue, const V& value)
+    {
+        amortizedCleanupIfNeeded();
+        return m_set.insertBefore(*static_cast<const T&>(beforeValue).weakPtrFactory().template createWeakPtr<T>(const_cast<U&>(beforeValue), assertionsPolicy).m_impl,
+            *static_cast<const T&>(value).weakPtrFactory().template createWeakPtr<T>(const_cast<V&>(value), assertionsPolicy).m_impl);
+    }
+
+    template <typename U>
+    AddResult insertBefore(iterator it, const U& value)
+    {
+        amortizedCleanupIfNeeded();
+        return m_set.insertBefore(it.m_position, *static_cast<const T&>(value).weakPtrFactory().template createWeakPtr<T>(const_cast<U&>(value), assertionsPolicy).m_impl);
+    }
+
+    template <typename U>
+    bool remove(const U& value)
+    {
+        amortizedCleanupIfNeeded();
+        auto& weakPtrImpl = value.weakPtrFactory().m_impl;
+        if (auto* pointer = weakPtrImpl.pointer(); pointer && *pointer)
+            return m_set.remove(*pointer);
+        return false;
+    }
+
+    bool remove(iterator it)
+    {
+        if (it == end())
+            return false;
+        auto result = m_set.remove(it.m_position);
+        amortizedCleanupIfNeeded();
+        return result;
+    }
+
+    void clear()
+    {
+        m_set.clear();
+        m_operationCountSinceLastCleanup = 0;
+    }
+
+    unsigned capacity() const { return m_set.capacity(); }
+
+    bool isEmptyIgnoringNullReferences() const
+    {
+        if (m_set.isEmpty())
+            return true;
+
+        auto onlyContainsNullReferences = begin() == end();
+        if (UNLIKELY(onlyContainsNullReferences))
+            const_cast<WeakListHashSet&>(*this).clear();
+        return onlyContainsNullReferences;
+    }
+
+    bool hasNullReferences() const
+    {
+        unsigned count = 0;
+        auto result = WTF::anyOf(m_set, [&] (auto& iterator) {
+            ++count;
+            return !iterator.get();
+        });
+        if (result)
+            amortizedCleanupIfNeeded(count);
```

Comment:
I think this should be `increaseOperationCountSinceLastCleanup(count);` for consistency with other weak containers now.

Reply:
Good point. Will fix.